### PR TITLE
[Quantization]support exclude operators while quantization

### PR DESCRIPTION
**Pull request by xinyu-intel**
## Description ##
Two functionally enhancement for quantization tool:
- support exclude operators while quantization.
- address #15796 , automatically exclude operator which is not registered with a compute function on the target device.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
@ZhennanQin @ciyongch please take a review :)

File: include/mxnet/c_api.h [hunk 1]
```diff
@@ -1909,16 +1909,22 @@ MXNET_DLL int MXSymbolInferTypePartial(SymbolHandle sym,
  * \brief Convert a symbol into a quantized symbol where FP32 operators are replaced with INT8
  * \param sym_handle symbol to be converted
  * \param ret_sym_handle quantized symbol result
- * \param num_excluded_symbols number of layers excluded from being quantized in the input symbol
- * \param excluded_symbols op names to be excluded from being quantized
+ * \param dev_type device type
+ * \param num_excluded_sym_names number of layers excluded from being quantized in the input symbol
+ * \param excluded_sym_names node names to be excluded from being quantized
+ * \param num_excluded_op_names number of operators excluded from being quantized in the input symbol
+ * \param excluded_op_names operator names to be excluded from being quantized
```
Comment (pengzhao-intel) [inline include/mxnet/c_api.h]: Is it possible to use one group of the parameter to implement two functionality in here?

File: src/operator/quantization/quantize_graph_pass.cc [hunk 2]
```diff
@@ -102,28 +102,57 @@ std::vector<NodeEntry> OfflineParams(std::vector<NodeEntry>&& outputs,
...
   if (op && quantized_op_map.count(op)) {
     bool need = true;
-    if (excluded_nodes.count(node->attrs.name)) {
+    // If the quantized node is not registered with a computation function, the node
+    // will be excluded automatically.
+    auto q_ptr = quantized_op_map[node->op()];
+    auto qnode = q_ptr(node->attrs);
+    if (!isRegistered(qnode, dev_type)) {
+      LOG(INFO) << "Neither FCompute nor FComputeEx registered, " << node->op()->name
+                << " excluded automatically.";
```
Comment (ZhennanQin) [inline src/operator/quantization/quantize_graph_pass.cc]: excluded  => is excluded

File: src/operator/quantization/quantize_graph_pass.cc [hunk 3]
```diff
@@ -102,28 +102,57 @@ std::vector<NodeEntry> OfflineParams(std::vector<NodeEntry>&& outputs,
...
+      } else if (!node->attrs.subgraphs.empty()) {
+        ExecType exec_type = fexec_type.count(op) ? fexec_type[op](node->attrs) : ExecType::kSync;
+        if (exec_type != ExecType::kSubgraphExec) {
+          // This is a fused subgraph node, try to match inner node.
+          CHECK_EQ(node->attrs.subgraphs.size(), 1);
+          auto subgraph_sym = node->attrs.subgraphs[0];
+          DFSVisit(subgraph_sym->outputs, [&](const nnvm::NodePtr& n) {
+            if (n->is_variable()) return;
+            if (excluded_nodes.count(n->attrs.name) ||
+                excluded_ops.count(node->op()->name)) {
```
Comment (ZhennanQin) [inline src/operator/quantization/quantize_graph_pass.cc]: Fused op is a new op, so we shouldn't check its inner node.

File: include/mxnet/c_api.h (hunk 1 above)
Comment (xinyu-intel) [reply include/mxnet/c_api.h]: some models may define layer names with specific style. so, it's not easy to group these two functions:(

File: python/mxnet/contrib/quantization.py [hunk 4]
```diff
@@ -758,11 +745,6 @@ def calib_graph(qsym, arg_params, aux_params, collector,
     """User-level API for calibrating a quantized model using a filled collector.
     The backend quantized operators are only enabled for Linux systems. Please do not run
     inference using the quantized models on Windows for now.
-    The quantization implementation adopts the TensorFlow's approach:
-    https://www.tensorflow.org/performance/quantization.
-    The calibration implementation borrows the idea of Nvidia's 8-bit Inference with TensorRT:
-    http://on-demand.gputechconf.com/gtc/2017/presentation/s7310-8-bit-inference-with-tensorrt.pdf
-    and adapts the method to MXNet.
```
Comment (ciyongch) [inline python/mxnet/contrib/quantization.py]: Keep this notes here for future reference.

File: python/mxnet/contrib/quantization.py [hunk 5]
```diff
@@ -821,16 +803,13 @@ def calib_graph(qsym, arg_params, aux_params, collector,
...
+                 calib_data=None, data_shapes=None, calib_mode='none',
+                 num_calib_examples=None, ctx=cpu(), logger=logging):
     """User-level API for Gluon users to generate a quantized SymbolBlock from a FP32 HybridBlock w/ or w/o calibration.
     The backend quantized operators are only enabled for Linux systems. Please do not run
     inference using the quantized models on Windows for now.
-    The quantization implementation adopts the TensorFlow's approach:
-    https://www.tensorflow.org/performance/quantization.
-    The calibration implementation borrows the idea of Nvidia's 8-bit Inference with TensorRT:
-    http://on-demand.gputechconf.com/gtc/2017/presentation/s7310-8-bit-inference-with-tensorrt.pdf
-    and adapts the method to MXNet.
```
Comment (ciyongch) [inline python/mxnet/contrib/quantization.py]: Keep this notes here for future reference.

File: tests/python/quantization/test_quantization.py [hunk 6]
```diff
@@ -678,6 +678,10 @@ def check_quantized_bn(data_shape, qdtype):
 
 @with_seed()
 def test_quantize_params():
+    if is_test_for_native_cpu():
+        print('skipped testing quantized_pooling for native cpu since it is not supported yet')
```
Comment (ciyongch) [inline tests/python/quantization/test_quantization.py]: test name? `testing quantize_params` ?

File: python/mxnet/contrib/quantization.py (hunk 5 above)
Comment (xinyu-intel) [reply python/mxnet/contrib/quantization.py]: file too long (>1000L):(

File: python/mxnet/contrib/quantization.py (hunk 4 above)
Comment (xinyu-intel) [reply python/mxnet/contrib/quantization.py]: file too long (>1000L):(

File: tests/python/quantization/test_quantization.py (hunk 6 above)
Comment (xinyu-intel) [reply tests/python/quantization/test_quantization.py]: okay

**Comment by pengzhao-intel** [comment]
@ZhennanQin @ciyongch please take a final review :)

File: src/operator/quantization/quantize_graph_pass.cc (hunk 3 above)
Comment (xinyu-intel) [reply src/operator/quantization/quantize_graph_pass.cc]: I found we cannot exclude fused conv layers when setting`excluded_op_names=['Convolution']`. Is it necessary to check the inner node here?

**Pull request merged**